flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractStateChangeLogger.java

```diff
@@ -0,0 +1,172 @@
+        this.stateChangelogWriter = checkNotNull(stateChangelogWriter);
+        this.keyContext = checkNotNull(keyContext);
+    }
+
+    @Override
+    public void valueUpdated(Value newValue, Ns ns) throws IOException {
+        if (newValue == null) {
+            valueCleared(ns);
+        } else {
+            log(SET, out -> serializeValue(newValue, out), ns);
+        }
+    }
+
+    @Override
+    public void valueUpdatedInternal(Value newValue, Ns ns) throws IOException {
+        if (newValue == null) {
+            valueCleared(ns);
+        } else {
+            log(SET_INTERNAL, out -> serializeValue(newValue, out), ns);
+        }
+    }
+
+    protected abstract void serializeValue(Value value, DataOutputViewStreamWrapper out)
+            throws IOException;
+
+    @Override
+    public void valueAdded(Value addedValue, Ns ns) throws IOException {
+        log(ADD, out -> serializeValue(addedValue, out), ns);
+    }
+
+    @Override
+    public void valueCleared(Ns ns) throws IOException {
+        log(CLEAR, out -> {}, ns);
+    }
+
+    @Override
+    public void valueElementAdded(
+            ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataSerializer, Ns ns)
+            throws IOException {
+        log(ADD_ELEMENT, dataSerializer, ns);
+    }
+
+    @Override
+    public void valueElementAddedOrUpdated(
+            ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataSerializer, Ns ns)
+            throws IOException {
+        log(ADD_OR_UPDATE_ELEMENT, dataSerializer, ns);
+    }
+
+    @Override
+    public void valueElementRemoved(
+            ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataSerializer, Ns ns)
+            throws IOException {
+        log(REMOVE_ELEMENT, dataSerializer, ns);
+    }
+
+    protected void log(
+            StateChangeOperation op,
+            ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataWriter,
+            Ns ns)
+            throws IOException {
+        // todo: log metadata (FLINK-22808)
+        stateChangelogWriter.append(
+                keyContext.getCurrentKeyGroupIndex(), serialize(op, ns, dataWriter));
+    }
+
+    private byte[] serialize(
+            StateChangeOperation op,
+            Ns ns,
+            ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataWriter)
+            throws IOException {
+        return serializeRaw(
+                wrapper -> {
+                    wrapper.writeByte(op.code);
+                    serializeScope(ns, wrapper);
+                    dataWriter.accept(wrapper);
+                });
+    }
+
+    protected abstract void serializeScope(Ns ns, DataOutputViewStreamWrapper out)
+            throws IOException;
+
+    private byte[] serializeRaw(
+            ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataWriter)
+            throws IOException {
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream();
```

Comment:
I believe you will add TODO for performance in https://github.com/apache/flink/pull/15420, right?

Reply:
Right (in 98d8eadd9fa4a52e6f8eaa7b62458a9d9cf76bc8).